### ISSUE TITLE
Feature request: proxy (with auth) support 

### ISSUE THREAD

**Issue by zdenop** [cppan/cppan#1]
It seams (Requesting dependency list Couldn't connect to server) that cppan windows client does not support access to internet via proxy.

**Comment by egorpugin** [comment]
Do you have a user:password pair? Or proxy bypasses every request.
I'm thinking about adding an option to cppan command.

**Comment by zdenop** [comment]
Yes I have. Something like curl is using would be nice:
`curl -O -v -U user:password -x proxy_address:proxy_port --url url`

**Comment by egorpugin** [comment]
I've added experimental support of proxies. Please, try:
1. Download master branch client https://cppan.org/client/cppan-master-win32-client.zip
2. Set 'proxy_host' and 'proxy_user' in %userprofile%/.cppan/.cppan file to your settings.
3. Try to run cppan for your project under proxy.

Example of c:\users\egor\.cppan\.cppan:
```
proxy_host: "http://myproxy:8080"
proxy_user: "myuser:mypasswd"
```

**Comment by zdenop** [comment]
I tried but it does not work for me:
```
Requesting dependency list
Downloading: pvt.cppan.demo.gif-5.1.2
Couldn't connect to server
```

**Comment by egorpugin** [comment]
Atleast it passes first request to server and fail during data download. I forget to add proxy settings there. Not I've fixed it. Please, re-download https://cppan.org/client/cppan-master-win32-client.zip and try again.

**Comment by zdenop** [comment]
now the output is different (so there is progress :-) ):
```
Requesting dependency list
Downloading: pvt.cppan.demo.gif-5.1.2
Failure when receiving data from the peer
```

**Comment by egorpugin** [comment]
Could you please try to download file directly via curl and post here your command line?
Because I don't have a proxy to check this.
Try to download a file via http and https:
```
https://cppan.org//data/pvt/c/cp/cppan/demo/gif/5.1.2.tar.gz
http://cppan.org//data/pvt/c/cp/cppan/demo/gif/5.1.2.tar.gz
```

**Comment by zdenop** [comment]
`curl -O -v -U zdenko.podobny:korela01 -x 10.204.10.2:3128 --url https://cppan.org//data/pvt/c/cp/cppan/demo/gif/5.1.2.tar.gz`

produced this:
```
* Hostname was NOT found in DNS cache
*   Trying 10.10.10.10...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 10.10.10.10 (10.10.10.10) port 3128 (#0)
* Establish HTTP proxy tunnel to cppan.org:443
* Proxy auth using Basic with user 'user'
> CONNECT cppan.org:443 HTTP/1.1
> Host: cppan.org:443
> Proxy-Authorization: Basic emRlbmtvLnBvZG9ibnk6a29yZWxhMDE=
> User-Agent: curl/7.35.0
> Proxy-Connection: Keep-Alive
>
< HTTP/1.0 200 Connection established
<
* Proxy replied OK to CONNECT request
* successfully set certificate verify locations:
*   CAfile: C:\Program Files\cURL\bin\curl-ca-bundle.crt
  CApath: none
* SSLv3, TLS handshake, Client hello (1):
} [data not shown]
* SSLv3, TLS handshake, Server hello (2):
{ [data not shown]
* SSLv3, TLS handshake, CERT (11):
{ [data not shown]
* SSLv3, TLS handshake, Server key exchange (12):
{ [data not shown]
* SSLv3, TLS handshake, Server finished (14):
{ [data not shown]
* SSLv3, TLS handshake, Client key exchange (16):
} [data not shown]
* SSLv3, TLS change cipher, Client hello (1):
} [data not shown]
* SSLv3, TLS handshake, Finished (20):
} [data not shown]
* SSLv3, TLS change cipher, Client hello (1):
{ [data not shown]
* SSLv3, TLS handshake, Finished (20):
{ [data not shown]
* SSL connection using ECDHE-RSA-AES256-GCM-SHA384
* Server certificate:
*        subject: CN=cppan.org
*        start date: 2016-05-13 09:51:00 GMT
*        expire date: 2016-08-11 09:51:00 GMT
*        subjectAltName: cppan.org matched
*        issuer: C=US; O=Let's Encrypt; CN=Let's Encrypt Authority X3
*        SSL certificate verify ok.
> GET //data/pvt/c/cp/cppan/demo/gif/5.1.2.tar.gz HTTP/1.1
> User-Agent: curl/7.35.0
> Host: cppan.org
> Accept: */*
>
< HTTP/1.1 301 Moved Permanently
* Server nginx/1.11.1 is not blacklisted
< Server: nginx/1.11.1
< Date: Fri, 01 Jul 2016 11:33:43 GMT
< Content-Type: text/html
< Content-Length: 185
< Connection: keep-alive
< Location: http://cppan.org//data/pvt/c/cp/cppan/demo/gif/5.1.2.tar.gz
<
{ [data not shown]
100   185  100   185    0     0    194      0 --:--:-- --:--:-- --:--:--   201
* Connection #0 to host 10.10.10.10 left intact
```

It created file `5.1.2.tar.gz`, but it does not content data but error message:
```
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.11.1</center>
</body>
</html>
```

**Comment by zdenop** [comment]
`curl -O -v -U user:password -x 10.10.10.10:3128 --url http://cppan.org//data/pvt/c/cp/cppan/demo/gif/5.1.2.tar.gz`

produced this:

```
* Hostname was NOT found in DNS cache
*   Trying 10.10.10.10...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 10.10.10.10 (10.10.10.10) port 3128 (#0)
* Proxy auth using Basic with user 'zdenko.podobny'
> GET http://cppan.org//data/pvt/c/cp/cppan/demo/gif/5.1.2.tar.gz HTTP/1.1
> Proxy-Authorization: Basic emRlbmtvLnBvZG9ibnk6a29yZWxhMDE=
> User-Agent: curl/7.35.0
> Host: cppan.org
> Accept: */*
> Proxy-Connection: Keep-Alive
>
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
< Server: nginx/1.11.1
< Date: Fri, 01 Jul 2016 11:32:05 GMT
< Content-Type: application/octet-stream
< Content-Length: 32105
< Last-Modified: Thu, 30 Jun 2016 13:36:22 GMT
< ETag: "57752056-7d69"
< Accept-Ranges: bytes
< X-Cache: MISS from mail.intime.sk
< X-Cache-Lookup: MISS from mail.intime.sk:3128
< Age: 431
< X-Cache: HIT from gw.intime.sk
< X-Cache-Lookup: HIT from gw.intime.sk:3128
< Via: 1.1 mail.intime.sk:3128 (squid/2.7.STABLE9), 1.0 gw.intime.sk:3128 (squid/2.7.STABLE3)
* HTTP/1.0 connection set to keep alive!
< Connection: keep-alive
* HTTP/1.0 proxy connection set to keep alive!
< Proxy-Connection: keep-alive
<
{ [data not shown]
100 32105  100 32105    0     0   105k      0 --:--:-- --:--:-- --:--:--  118k
* Connection #0 to host 10.10.10.10 left intact
```
It created file `5.1.2.tar.gz`, that contains data.

**Comment by egorpugin** [comment]
You should use '-L' option: follow relocations.

**Comment by egorpugin** [comment]
I'll look more closely at this issue in 2-3 days. Sorry for the inconvenience.

**Comment by egorpugin** [comment]
Ok, after fixing stupid typos, proxy support should be fine.
I've also reorganized settings. Please, rewrite old 'proxy_host' and 'proxy_user' as listed below. (And re-download master client https://cppan.org/client/cppan-master-win32-client.zip.)
```
proxy:
  host: 192.168.191.129:3128
  user: proxy_test:123456
```

Please, close the ticket if things are ok for you.

**Comment by zdenop** [comment]
I did changes but the result is:
```
Requesting dependency list
Failure when receiving data from the peer
```

**Comment by egorpugin** [comment]
It's strange, because I set up a proxy locally and it works with user+password.
Did you specify your auth correctly?

**Comment by zdenop** [comment]
The current master version work for me (with option --ignore-ssl-checks)

**Comment by egorpugin** [comment]
I'm updating curl to the latest 7.51.0 version which fixed (somewhere between 7.49.0 and 7.50.0) https://github.com/curl/curl/issues/840
Also I'm removing my workaround. Updates will go live with the next client update.
